### PR TITLE
Create proposal for a default type operator analogous to function application

### PR DESCRIPTION
**Pull request by habibalamin**

### CODE REVIEW

File: proposals/0000-dollar-type-operator-as-default.rst [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+Dollar type operator as default
+===============================
+
+Define the dollar type operator (`type a $ b = a b`) as a default when using
+the `TypeOperators` extension.
+
+Motivation
+----------
+
+Though, I have admittedly only used the `TypeOperators` extension once, I
```
Comment (bgamari) [inline proposals/0000-dollar-type-operator-as-default.rst]: As a matter of style, personal anecdotes like this are generally best avoided as they add length yet don't help the reader understand the proposal.

File: proposals/0000-dollar-type-operator-as-default.rst [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+think that the main reason it is used is to have the function application
+operator available in type signatures. That's what I used it for.
+
+Proposed Change
+---------------
+
+When using the `TypeOperators` language extension, we should be able
```
Comment (bgamari) [inline proposals/0000-dollar-type-operator-as-default.rst]: Inline code elements are demarcated by double-ticks in ReST, not single-ticks like in Markdown

File: proposals/0000-dollar-type-operator-as-default.rst [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+to `value :: IO $ Maybe Int` rather than `value :: IO (Maybe Int)`,
+without having to define `type a $ b = a b`. It should be defined by
+default when I use the language extension.
+
+Drawbacks
+---------
+
+It forces us to start thinking about a Prelude-like default operators
+for the `TypeOperators` extension. It adds bloat to the extension and
+assigns it more than one responsibility (violating the SRP).
```
Comment (bgamari) [inline proposals/0000-dollar-type-operator-as-default.rst]: What is SRP?

**Comment by bgamari** [comment]
Thanks for your proposal! I've left a few points inline.

On the whole I'm not sure I like the idea of using language extensions to affect the visibility of `Prelude` definitions. Why not simply propose to add `type ($) a b = a b` (with the appropriate fixity) somewhere in `base` (and perhaps `Prelude`)? This would not be a GHC-specific proposal and would likely go through the Core Libraries Committee.

File: proposals/0000-dollar-type-operator-as-default.rst (hunk 3 above)
Comment (habibalamin) [inline proposals/0000-dollar-type-operator-as-default.rst]: Single Responsibility Principle. I'll spell it out.

**Comment by habibalamin** [comment]
Okay, I'll see if I can get them to consider it instead. Thank you!

**Comment by vagarenko** [comment]
It's not that simple.
`type ($) a b = a b` will work for `IO $ Maybe a` but will fail for multiple operators like `IO $ Maybe $ Either a b`.
See this blog post:
https://typesandkinds.wordpress.com/2013/04/01/defunctionalization-for-the-win/

**Comment by goldfirere** [comment]
I don't follow @vagarenko 's complaint. What fails about `IO $ Maybe $ Either a b`, as long as `$` has the same right-fixity as it does at the term level?

**Comment by phadej** [comment]
FWIW, I have 
```hs
type (:$) (f :: k -> l) (x :: k) = f x
infixr 0 :$
```
in our company prelude. It works fine.

**Pull request closed without merging**